### PR TITLE
Debug: Forward the command line debug environment variable to the browser

### PR DESCRIPTION
**Pull request by jordwest**
#### Why?
 - Don't lose the debug filter when localStorage is cleared (this happens every time the current user changes)
 - Set both the console debug message filter **and** the browser debug filter at the same time.
 - No more manually setting `debug` in localStorage each time you work on a different feature.
 - Capture boot/initialization debug messages from the first boot.

#### Changes

This change allows setting the debug output filter at the command line to have it passed to the browser:
```sh
DEBUG=calypso:* make run
```

Then get this in the browser:
screen shot 2016-03-02 at 1 01 26 pm

This will be set on each page request, avoiding the need to manually set the localStorage value each time the user changes/logs in/logs out.

### CODE REVIEW

**Comment by klimeryk** [comment]
Nice improvements and good idea!
The only gripe I have with this implementation is the confusing persistence of `DEBUG`'s value. If you run `DEBUG=calypso:* make run` once, and then restart using just `make run` (without changing user, clearing `localStorage`, etc). you should get the same results as with `DEBUG=calypso:* make run`, right? Because `DEBUG`'s value is persisted in `localStorage`. But like you said, changing the current user scrubs that. This all seems a bit confusing to me.
I'd propose to move the debug filter's value to a file, where you can update it as needed (and you don't have to specify it from the command line). Maybe it can be in `/config` files? For development, use `calypso:*` as a good starting point (for developers starting with Calypso that seems like a sane default, if a _bit_ verbose) and let everyone override it to whatever part of Calypso they are working on using the awesome [`/config/*.local.json` cascading config files feature](https://calypsop2.wordpress.com/2016/02/16/casading-configs/)?

**Comment by jordwest** [comment]
@klimeryk thanks for the review. I agree that the persistence is confusing.

Great idea with the config files, I also like the idea of having a default of max verbosity.

There are a couple of benefits I can see with environment variables:
 - The messages will be consistent on both server side (dev server stdout) and in the browser
 - It could be possible to set up local git hooks to automatically set the environment variable when switching to specific branches

Maybe we could go with a hybrid approach - set a default of 'calypso:*' in `development.json`, then allow the user to override that with `DEBUG=xyz make run`. That way if no env var is provided, it will reset to the default of 'calypso:*'. This would also match the behaviour of feature flags. What do you think?

**Comment by klimeryk** [comment]
Definitely! I like the hybrid approach :+1: I was thinking the same thing - `DEBUG` should work like `ENABLE_FEATURE` for features. To temporarily change the behavior - till the next restart. For persistent changes use config files.

> The messages will be consistent on both server side (dev server stdout) and in the browser

Can we not also set the `debug` level on the server side based on the configs? We're already, for example, [parsing the config files on the server](https://github.com/Automattic/wp-calypso/blob/master/server/config/parser.js), so it should be possible? Or am I missing something here?

> It could be possible to set up local git hooks to automatically set the environment variable when switching to specific branches

Hmm, I'm not sure I see the use case here (then again, I don't work on too many features concurrently). Can you provide an example? :)

**Comment by jordwest** [comment]
Hybrid approach implemented and working well :)

> Can we not also set the debug level on the server side based on the configs? We're already, for example, parsing the config files on the server, so it should be possible? Or am I missing something here?

You're completely right, I've gone ahead and implemented that. I can confirm that setting the config to `"debug_filter": "*"` shows a ridiculous number of debug messages on both the server and browser, and starting the server with `DEBUG= make run` will prevent all messages.

> Hmm, I'm not sure I see the use case here (then again, I don't work on too many features concurrently). Can you provide an example? :)

I'll admit I haven't tried it, but the idea was something like this:

1. Add a `post-checkout` git hook.
2. When switching to a branch containing `support-user` in the name, set `DEBUG=calypso:support-user`
3. When switching to a branch containing `exporter`, set `DEBUG=calypso:exporter`

This would all be local of course and manually updated, but it would support both existing and future PRs/branches.

**Comment by klimeryk** [comment]
Sweet! I very much like how this turned out :+1:
As for the `post-checkout` idea - I see it's merit, but it could also lead to some confusion. We could, of course, add some `echo`s to let the user know that `DEBUG` has been automagically set based on the name of the branch. If we only `export DEBUG=calypso:your-feature` in the `post-checkout` hook, then when the developer changes terminals, the value of `DEBUG` set that way won't be persisted. We could modify the value in `development.local.json`, but that we would have to add a check for existing values, etc. Seems too complicated for the small (IMHO) gains. Either way, I think implementation of such a hook can be left for another PR, maybe after a discussion here (on a separate issue) or on a P2.

As for the PR itself - you are not taking into account a situation when `debug_filter` is not present in the config :) So `wp-calypso` or `production` builds are failing :disappointed: Adding `"debug_filter": ""` to `/config/_shared.json` should fix this. You could also `catch` the error thrown by `config( 'debug_filter' )` when the key does not exist and provide a sane default then. But it seems the convention is that the non-`features` values in configs are mandatory and should always be present.

**Comment by jordwest** [comment]
Oh sorry, I should have explained better. I wasn't suggesting pushing the hook to the remote repo, I was thinking the git hook could be added locally (manually). After having a go I could publish a gist for example, and contributors could add it to their local copy/customize it at their leisure, but only for those who intentionally choose to; as you mentioned it could be too much cause for confusion to have something like that as a default.

Thanks for helping get this pushed along! Would you be comfortable to give it a :+1: in it's current state?

**Comment by klimeryk** [comment]
Ah, that makes way more sense, than what I was suggesting :laughing: 
In that case, a well deserved :+1:

**Comment by nylen** [comment]
Starting the app with consistent debug settings is a good idea, but preserving the ability to manipulate `localStorage.debug` in the console (and **remembering its value across page loads**) is an important feature when you're working on a particular component or area of the app.  It *looks* like this PR will reset `localStorage.debug` on each page refresh, **and** require a server restart to make changes to the debug settings, is that correct?

**Comment by jordwest** [comment]
Correct, this effectively means we would no longer set the filter manually in the browser. The standard way to change the filter would then be via the config or the `DEBUG` environment var and you would need to restart the server each time the debug filter is changed.

In a sense makes the debug filter more persistent across page loads (it's not lost on login/logout/user switching), at the cost of having to restart the server each time the filter is changed.

We could continue to change the `localStorage.debug` directly, but the question then is: when do we update that with the value set in the configuration?

**Comment by kwight** [comment]
I feel the same as @nylen : quickly changing between debug settings is useful when trying to narrow something down. That said, I admit using the console filter seems to work just as well.

**Comment by klimeryk** [comment]
I think it's just a matter of different habits. With the old approach, when you changed the debug filter in local storage, you still had to refresh to see the change, right? That seems wrong to me - that would be like expecting the browser to save the changes I've made to the CSS rules using the developer tools. It's good for a quick test, but if you want to persist it, you should modify the code directly.
Similarly, the debug filter is something that controls the code, so it should be defined there, not in the browser where it's volatile. As @kwight suggested, if you need to narrow down something quickly, you can use the built in filtering of developer tools. Is there a use case that cannot be handled by such filtering? (assuming you're already on a debug filter that contains the message you are looking for)
I'd really love to know if that's the case - I usually just keeping to the `calypso:*` filter, so I might be missing some use cases.

**Comment by jordwest** [comment]
One potential solution is that we could keep track of the server debug value in a separate localstorage value (eg, `server_debug`), then only copy it to the `localStorage.debug` if either:

 - `localStorage.debug` is undefined; or,
 - `localStorage.debug` === `localStorage.server_debug` before loading the new value from server

That way, when `localStorage.debug` is changed from the browser, it won't be automatically updated with the server's debug filter. Clearing the store would then switch back to the server's debug filter.

It could all get a bit confusing, though, having three different places to set the debug value.

**Comment by nylen** [comment]
I was going to make a similar, but more simple, suggestion.  Only reset the `localStorage.debug` filter in the following two cases:

- On the first page load from a new `make run`, set it to the server debug value.
- If `localStorage.debug` is empty on a page load, set it to the server debug value.

File: server/pages/index.jade [hunk 1]
```diff
@@ -117,6 +117,11 @@ html(lang=lang, dir=isRTL ? 'rtl' : 'ltr', class=isFluidWidth ? 'is-fluid-width'
 			script(type='text/javascript')!='var currentUser = ' + sanitize.jsonStringifyForHtml( user )
 		if app
 			script(type='text/javascript')!='var app = ' + sanitize.jsonStringifyForHtml( app )
+		if 'development' === env
+			script.
+				if ( window.app && window.app.debugFilter && localStorage ) {
```
Comment (mcsf) [inline server/pages/index.jade]: Should this line read `&& window.localStorage`?

**Pull request closed without merging**